internal/coord/routing/crawl.go

```diff
@@ -0,0 +1,357 @@
+		return nil, err
+	}
+
+	c := &Crawl[K, N]{
+		self:  self,
+		cfg:   *cfg,
+		cplFn: cplFn,
+		info:  nil,
+	}
+
+	return c, nil
+}
+
+func (c *Crawl[K, N]) Advance(ctx context.Context, ev CrawlEvent) (out CrawlState) {
+	_, span := c.cfg.Tracer.Start(ctx, "Crawl.Advance", trace.WithAttributes(tele.AttrInEvent(ev)))
+	c.setMapSizes(span, "before")
+	defer func() {
+		c.setMapSizes(span, "after")
+		span.SetAttributes(tele.AttrOutEvent(out))
+		span.End()
+	}()
+
+	switch tev := ev.(type) {
+	case *EventCrawlStart[K, N]:
+		if c.info != nil {
+			break // query in progress, pretend it was a poll
+		}
+
+		span.SetAttributes(attribute.Int("seed", len(tev.Seed)))
+
+		ci := &crawlInformation[K, N]{
+			todo:    []crawlJob[K, N]{},
+			cpls:    map[string]int{},
+			waiting: map[string]N{},
+			success: map[string]N{},
+			failed:  map[string]N{},
+			errors:  map[string]error{},
+		}
+
+		for _, node := range tev.Seed {
+			// exclude self from closest nodes
+			if key.Equal(node.Key(), c.self.Key()) {
+				continue
+			}
+
+			for j := 0; j < c.cfg.MaxCPL; j++ {
+				target, err := c.cplFn(node.Key(), j)
+				if err != nil {
+					// return nil, fmt.Errorf("generate cpl: %w", err)
+					// TODO: log
+					continue
+				}
+
+				job := crawlJob[K, N]{
+					node:   node,
+					target: target.Key(),
+				}
+
+				ci.cpls[job.mapKey()] = j
+				ci.todo = append(ci.todo, job)
+			}
+		}
+
+		c.info = ci
+
+	case *EventCrawlNodeResponse[K, N]:
+		span.SetAttributes(attribute.Int("closer_nodes", len(tev.CloserNodes)))
+
+		if c.info == nil {
+			return &StateCrawlIdle{}
+		}
+
+		job := crawlJob[K, N]{
+			node:   tev.NodeID,
+			target: tev.Target,
+		}
+
+		mapKey := job.mapKey()
+		if _, found := c.info.waiting[mapKey]; !found {
+			break
+		}
+
+		delete(c.info.waiting, mapKey)
+		c.info.success[mapKey] = tev.NodeID
+
+		for _, node := range tev.CloserNodes {
+			for i := 0; i < c.cfg.MaxCPL; i++ {
+				target, err := c.cplFn(node.Key(), i)
+				if err != nil {
+					// TODO: log
+					continue
+				}
+
+				newJob := crawlJob[K, N]{
+					node:   node,
+					target: target.Key(),
+				}
+
+				newMapKey := newJob.mapKey()
+				if _, found := c.info.cpls[newMapKey]; found {
+					continue
+				}
+
+				c.info.cpls[newMapKey] = i
+				c.info.todo = append(c.info.todo, newJob)
+			}
+		}
+
+	case *EventCrawlNodeFailure[K, N]:
+		if c.info == nil {
+			return &StateCrawlIdle{}
+		}
+
+		job := crawlJob[K, N]{
+			node:   tev.NodeID,
+			target: tev.Target,
+		}
+
+		mapKey := job.mapKey()
+		if _, found := c.info.waiting[mapKey]; !found {
+			break
+		}
+		span.RecordError(tev.Error)
+
+		delete(c.info.waiting, mapKey)
+		c.info.failed[mapKey] = tev.NodeID
+		c.info.errors[mapKey] = tev.Error
+
+	case *EventCrawlPoll:
+		// no event to process
+	default:
+		panic(fmt.Sprintf("unexpected event: %T", tev))
+	}
+
+	if c.info == nil {
+		return &StateCrawlIdle{}
+	}
+
+	if len(c.info.waiting) >= c.cfg.MaxCPL*c.cfg.Concurrency {
```

Comment:
```suggestion
	if len(c.info.waiting) >= c.cfg.Concurrency {
```
Concurrency is the maximum number of concurrent requests, but the original code is sending 16 times as many

Reply:
Sending 16 requests is intended. Each request contains a different target key for which we want to know the 20 closest nodes that the other peer knows. This is the strategy for effectively fetching the entire routing table of a remote peer.

Reply:
Concurrency should be the maximum number of in-flight requests. The 16 is irrelevant here, since we are checking how many are currently in-flight. As it stands if the user specifies concurrency of 200 then they will actually end up with 3200 concurrent requests.